ffs/scheduler/scheduler.go

```diff
@@ -422,7 +428,29 @@ func (s *Scheduler) executeQueuedStorage(j ffs.StorageJob) {
 
 	// Execute
 	s.l.Log(ctx, "Executing job %s...", j.ID)
-	info, dealErrors, err := s.executeStorage(ctx, a, j)
+	dealUpdates := make(chan ffs.FilStorage)
+	go func() {
+		for update := range dealUpdates {
+			s.jobStatusCacheLock.Lock()
+			log.Infof("ZZZ -- API: %v, Miner: %v, Status: %v", a.APIID, update.Miner, storagemarket.DealStates[update.Status])
+			_, ok := s.jobStatusCache[j.APIID]
+			if !ok {
+				s.jobStatusCache[j.APIID] = map[cid.Cid]map[cid.Cid]ffs.FilStorage{}
+			}
+			_, ok = s.jobStatusCache[j.APIID][j.Cid]
+			if !ok {
+				s.jobStatusCache[j.APIID][j.Cid] = map[cid.Cid]ffs.FilStorage{}
+			}
+			s.jobStatusCache[j.APIID][j.Cid][update.ProposalCid] = update
+			log.Infof("\n%v", s.jobStatusCache)
+			s.jobStatusCacheLock.Unlock()
+		}
+		s.jobStatusCacheLock.Lock()
+		delete(s.jobStatusCache[j.APIID], j.Cid)
+		s.jobStatusCacheLock.Unlock()
+		log.Info("ZZZ -- Done receiving updates.")
+	}()
+	info, dealErrors, err := s.executeStorage(ctx, a, j, dealUpdates)
```

Comment:
Here's the big change: We pass in an channel of `ffs.FilStorage` here. It internally gets passed down into the lower level plumbing that executes cold storage and waits for deals to become final. That lower level code now publishes events on this channel so we can subscribe to them here.

Reply:
The other key thing here:

We now have events about how individual deals are progressing for a specific FFS instance and data cid here in `Scheduler`. These events are available, potentially, in the same scope as events that trigger updates to listeners of watching Jobs... Currently, `Scheduler` calls methods on `sjstore`, and those calls trigger updates to Job watchers. We could lift those event triggers and job subscriber management up out our `sjstore` to `Scheduler` and that would give us the opportunity to send job updates triggered by both `sjstore` methods calls as well as related deal events.

Reply:
Cool, looks like might be useful to also provide structured events.